File: src/leetcode/python/two_pointers/three_sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:

        result = []

        # Sort the numbers first so we could apply 2 pointers technique
        nums.sort()

        # We are looking for a triple of numbers a,b,c to
        # add up to 0.

        # Terminal conditions:
        # 1. If sorted and the current a value is > 0, there is no way it adds 2 more number and still = 0.
        # The triplets can never start with the current number

        for i, a in enumerate(nums):
            if a > 0:
                break

            # Duplicate happens, we skip it
            if i > 0 and a == nums[i - 1]:
                continue

            # Suppose we have a already, we just have to find b and c using 2 pointers, similar to two sum ii

            l = i + 1
            r = len(nums) - 1

            while l < r:
                threesum = a + nums[l] + nums[r]
                if threesum > 0:
                    r -= 1
                elif threesum < 0:
                    l += 1
                else:
                    result.append([a, nums[l], nums[r]])
                    # now we need to move on to other possibilities within l and r
                    l += 1
                    r -= 1
                    while l < r and nums[l] == nums[l - 1]:
                        l += 1
        return result


if __name__ == "__main__":
    sol = Solution()

File: src/leetcode/julia/two_pointers/threeSum.jl
function threeSum(nums::Vector{Int})::Vector{Vector{Int}}

    res = Vector{Vector{Int}}()
    sort!(nums) # sort the array in order to apply two pointers

    # check early termination
    if length(nums) < 3
        return res
    end

    for (i, a) in enumerate(nums)
        if a > 0
            break # the remaining elements are all positive, can no longer add up to 0
        end

        if i > 1 && a == nums[i - 1]
            continue # skip duplicates
        end
    

        l , r = i + 1, length(nums) # Julia's index starts from 1, so its length instead of length - 1

        while l < r
            threesum = a + nums[l] + nums[r]

            if threesum > 0 
                r -= 1
            elseif threesum < 0
                l += 1
            else 
                push!(res,[a, nums[l], nums[r]])
                l += 1
                r -= 1
                while nums[l] == nums[l - 1] && l < r # julia uses && instead of and
                    l += 1
                end
            end
        end
        end
    return res
end